# Семинар: знакомство с NumPy. Задачи
[(**Num**eric **Py**thon)](http://www.numpy.org/)


### Дополнительные материалы

Для самых любознательных и тех, кто хочет порешать задачи, есть сборник "100 заданий по Numpy": https://www.machinelearningplus.com/python/101-numpy-exercises-python/

Для тех, кому нужно руководство по матричным операциям в Numpy: https://www.programiz.com/python-programming/matrix

**numpy reference:** https://numpy.org/doc/stable/reference/index.html

In [ ]:
import numpy as np

## Задания для самостоятельного решения

#### Часть 1

1. Развернуть одномерный массив (сделать так, чтобы его элементы шли в обратном порядке).
2. Найти максимальный нечетный элемент в массиве.
3. Замените все нечетные элементы массива на ваше любимое число.
4. Создайте массив первых n нечетных чисел, записанных в порядке убывания. Например, если `n=5`, то ответом будет `array([9, 7, 5, 3, 1])`. *Функции, которые могут пригодиться при решении: `.arange()`*
5. Вычислите самое близкое и самое дальнее числа к данному в рассматриваемом массиве чисел. Например, если на вход поступают массив `array([0, 1, 2, 3, 4])` и число 1.33, то ответом будет `(1, 4)`. _Функции, которые могут пригодиться при решении: `.abs()`, `.argmax()`, `.argmin()`_

1. Развернуть одномерный массив (сделать так, чтобы его элементы шли в обратном порядке).

In [ ]:
#Задача 1.1
#код:
a = np.array([1, 2, 3, 1, 0])
a[::-1]

array([0, 1, 3, 2, 1])

2. Найти максимальный нечетный элемент в массиве.

In [ ]:
#Задача 1.2
#код:
print(a)
max(a[np.where(~(a % 2 == 0))]) # == max(a[np.where((a % 2 != 0))])


[1 2 3 1 0]


3

3. Замените все нечетные элементы массива на ваше любимое число.

In [ ]:
#Задача 1.3
#код:

np.where(a % 2 != 0, 8, a)


array([8, 2, 8, 8, 0])

4. Создайте массив первых n нечетных чисел, записанных в порядке убывания. Например, если `n=5`, то ответом будет `array([9, 7, 5, 3, 1])`. *Функции, которые могут пригодиться при решении: `.arange()`*

In [ ]:
#Задача 1.4
#код:
n = 6
m = np.arange(1, n*2, 2)
m[::-1]

array([11,  9,  7,  5,  3,  1])

5. Вычислите самое близкое и самое дальнее числа к данному в рассматриваемом массиве чисел. Например, если на вход поступают массив `array([0, 1, 2, 3, 4])` и число 1.33, то ответом будет `(1, 4)`. _Функции, которые могут пригодиться при решении: `.abs()`, `.argmax()`, `.argmin()`_

In [ ]:
#Задача 1.5
#код:
a = np.array([0, 1, 2, 3, 4])
n = float(input())
np.round(n)
min_close = max(a[np.where((a <= np.round(n)))])
print(min_close, a.argmax())

1.33
1 4


### Задания для самостоятельного решения

#### Часть 2. В этом разделе использовать только методы numpy. Не использовать циклы!  
#### Каждая задача в этом разделе решается одной-двумя строками кода. Нужно просто подумать или найти подходящий метод в numpy

Везде, где встречаются массивы или матрицы, подразумевается, что это `numpy.array`

#### Задача 2.1

Создайте одномерный массив чисел от 0 до 8.

Сделайте это двумя способами: с помощью метода `arange` из numpy и с помощью преобразования списка (list) в numpy array

In [ ]:
#Код:
a = np.arange(0, 9)
b = np.array(list(range(0, 9)))
c = np.array([0, 1, 2, 3,4, 5, 6, 7, 8])
print(a)
print(np.array(b))
print(c)

[0 1 2 3 4 5 6 7 8]
[0 1 2 3 4 5 6 7 8]
[0 1 2 3 4 5 6 7 8]


#### Тесты:

Выходные данные: `array([0, 1, 2, 3, 4, 5, 6, 7, 8])`


#### Задача 2.2

Вам дан numpy array под названием ```test_array```. Выведите положительные элементы этого массива

*Подсказка:* здесь можно также использовать булевы массивы в качестве булевой маски.

In [ ]:
#Код:
test_array = np.array([ 0, -1, 5, -3, 7, -1, 6, -1, 8, -1])
test_array[test_array  > 0]

array([5, 7, 6, 8])

#### Тесты:  
Входные данные:   
`test_array = np.array([ 0, -1,  5, -3,  7, -1,  6, -1,  8, -1])`  
Выходные данные:   
`array([5, 7, 6, 8])`

#### Задача 2.3

Замените отрицательные элементы из массива ```test_array``` на 0. Подсказка: вам пригодится функция ```np.where```

In [ ]:
#Код:
test_array = np.array([ 0, -1, 5, -3, 7, -1, 6, -1, 8, -1])
np.where(test_array < 0, 0, test_array)


array([0, 0, 5, 0, 7, 0, 6, 0, 8, 0])

#### Тесты:  
Входные данные:   
`test_array = np.array([ 0, -1,  5, -3,  7, -1,  6, -1,  8, -1])`  
Выходные данные:   
`array([0, 0, 5, 0, 7, 0, 6, 0, 8, 0])`

#### Задача 2.4

Вам дан массив из 10 чисел. С помощью одной функции в numpy найдите позицию максимального элемента в заданном массиве

In [ ]:
#Код:
array = np.array([6, 8, 4, 2, 11, -3, 7, 22, -20, -30])
np.where(array == max(array))[0][0] # 1-ый вариант
np.argmax(array) # 2-ой вариант


7

#### Тесты:  
Входные данные:   
`array = np.array([6, 8, 4, 2, 11, -3, 7, 22, -20, -30])`  
Выходные данные:   
`7`

#### Задача 2.5

Вам дан словарь, в котором подсчитаны частоты женских имен в выборке из 26 новорожденных девочек.

In [ ]:
names = dict()
names['Катя'] = 8
names["Maша"] = 5
names["Вероника"] = 2
names["Ярослава"] = 1
names["Александра"] = 10

dict_value = names.values()
dict_value_array = np.array(list(dict_value))
result = dict_value_array/sum(dict_value_array)
print(result, sum(result))

[0.30769231 0.19230769 0.07692308 0.03846154 0.38461538] 1.0


Посчитайте процентные доли имен с помощью нормирования вектора значений.

*Подсказка:* просто возьмите этот вектор значений и поделите его на сумму элементов в нем

#### Тесты:

Выходные данные:  
`array([0.30769231, 0.19230769, 0.07692308, 0.03846154, 0.38461538])`


Примечание: мы только что произвели операцию **нормирования вектора** (теперь суммa его элементов 1), а получившиеся элементы можно интерпретировать как **вероятности** (в примере выше это веростности того, что случайно взятая новорожденная девочка будет иметь какое-то конкретное имя)

#### Задача 2.6

Вам дан массив ```arr```. Выведите все его элементы, которые строго больше 5, и строго меньше 9

*Подсказка:* здесь можно использовать булевы массивы в качестве булевой маски.

In [ ]:
#Код:
arr = np.array([7, 2, 10, 2, 7, 4, 9, 4, 9, 8])
arr = arr[arr > 5]
arr = arr[arr < 9]
arr

array([7, 7, 8])

#### Тесты:

Входные данные:  
`arr = np.array([7, 2, 10, 2, 7, 4, 9, 4, 9, 8])`  
Выходные данные:     
`array([7, 7, 8])`


#### Двумерные массивы

#### Задача 2.7

Создайте методом из numpy одномерный массив из 12 элементов от 0 до 11 и сделайте из него матрицу размером 3х4 (вам пригодится функция ```np.reshape```)

In [ ]:
#Код:
m = np.arange(0, 12)
m.reshape(3,4)

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

#### Задача 2.8

Соедините вертикально и горизонтально две созданные матрицы a и b.

In [ ]:
a = np.arange(8).reshape(2,4)
b = np.ones(8).reshape(2,4)
print(a)
print()
print(b)
print()
print(np.hstack((a, b)))
print()
print(np.vstack((a, b)))

[[0 1 2 3]
 [4 5 6 7]]

[[1. 1. 1. 1.]
 [1. 1. 1. 1.]]

[[0. 1. 2. 3. 1. 1. 1. 1.]
 [4. 5. 6. 7. 1. 1. 1. 1.]]

[[0. 1. 2. 3.]
 [4. 5. 6. 7.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]]


#### Задача 2.9

Сохраните в отдельную переменную второй столбец (столбец с индексом 1) из данного вам двумерного массива.

Обратите внимение, что при вырезании столбца из матрицы, у вас получится одномерный numpy-array, а не двумерная матрица с одним столбцом!

In [ ]:
arr = np.arange(12).reshape(3,4)
print(arr)
print()
arr[:, 1]

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]



array([1, 5, 9])

#### Задача 2.10

Создайте массив из случайных чисел размера 3 на 4 (3 строки и 4 столбца) и сохраните его в переменную ```random_array```. Найдите среднее, медиану, стандартное отклонение созданной матрицы ```random_array``` по строкам и по столбцам. По одной строке на среднее, медиану, и стандартное отклонение для операций по строкам и по столбцам - всего 6 строк кода

In [ ]:
#Код:
random_array = np.round(np.random.rand(3, 4) * 10)
print(random_array)
print(f"Среднее значение: {np.mean(random_array)}")
print(f"Медианное значение: {np.median(random_array)}")
print(f"Стандартное отклонение: {np.std(random_array)}")
print(f"Стандартное отклонение по стоблцам: {np.std(random_array, axis = 0)}")
print(f"Стандартное отклонение по строкам: {np.std(random_array, axis = 1)}")




[[7. 6. 7. 7.]
 [7. 8. 2. 9.]
 [6. 6. 8. 0.]]
Среднее значение: 6.083333333333333
Медианное значение: 7.0
Стандартное отклонение: 2.465033242958173
Стандартное отклонение по стоблцам: [0.47140452 0.94280904 2.62466929 3.8586123 ]
Стандартное отклонение по строкам: [0.4330127 2.6925824 3.       ]


#### Задача 2.11

Найдите средние значения по строкам у созданной в задании 2.10 матрицы ```random_array```. Вычтите полученные средния значения из соответствующих строк начальной матрицы ```random_array```. Проверьте, что после вычитания средних среднее по каждой строке стало равно 0.

In [ ]:
#Код:
mean_line = np.mean(random_array, axis = 1)
print(f"Среднее значение по строкам: {mean_line}")
print()
print(f"Превратили среднее значение в вектор: \n {mean_line.reshape(3, 1)}")
print()
difference =  random_array - mean_line.reshape(3, 1)
print(f"Исходная матрица: \n {random_array} \n")
print(f"Матрица, которая получилась при вычитании среднего по строке из каждого элемента в строке: \n {difference}")
print("Тпеперь снова считаем среднее по строке в полученной матрице")
np.mean(difference, axis = 1)

Среднее значение по строкам: [2.75 5.25 1.25]

Превратили среднее значение в вектор: 
 [[2.75]
 [5.25]
 [1.25]]

Исходная матрица: 
 [[1. 3. 2. 5.]
 [8. 7. 3. 3.]
 [1. 2. 0. 2.]] 

Матрица, которая получилась при вычитании среднего по строке из каждого элемента в строке: 
 [[-1.75  0.25 -0.75  2.25]
 [ 2.75  1.75 -2.25 -2.25]
 [-0.25  0.75 -1.25  0.75]]
Тпеперь снова считаем среднее по строке в полученной матрице


array([0., 0., 0.])

#### Задача 2.12
Пусть мы имеем 2 массива (заданы ниже в коде). Требуется создать массив, который получает результат поэлементно True, если четные элементы 1го массива больше 2го и True, если нечетные элементы 2го больше элементов первого. В противном случае возвращать False.

Реализовать задачу с помощью numpy

Пример массивов и результата:

a = [7, 3, 5, 9, 2, 5, 4, 8, 9, 0]

b = [2, 8, 6, 7, 1, 3, 2, 9, 5, 1]

res = [ True, True, False, False, True, False, True, True, True, True]

Вы можете сделать задачу с помощью слайсинга или комбинации функций np.arange и np.where

In [ ]:
#Код:
a = np.array([7, 3, 5, 9, 2, 5, 4, 8, 9, 0])
b = np.array ([2, 8, 6, 7, 1, 3, 2, 9, 5, 1])

res = np.zeros(shape = a.shape, dtype = bool)
# print(a[::2] > b[::2])
# print(a[1::2] < b[1::2])
res[::2] = a[::2] > b[::2]
res[1::2] = a[1::2] < b[1::2]
res

array([ True,  True, False, False,  True, False,  True,  True,  True,
        True])

#### Задача 2.13
Создайте массив размера 10 со значениями от 0 до 1 (интервал между значениями равномерный), не включая концы. См np.linspace + слайсинг

In [ ]:

#Код:
np.linspace(0, 1, 12)[1:-1]

array([0.09090909, 0.18181818, 0.27272727, 0.36363636, 0.45454545,
       0.54545455, 0.63636364, 0.72727273, 0.81818182, 0.90909091])

#### Задача 2.14
Создайте случайный вектор размера 10, и замените в нем максимальное значение на 0

In [ ]:
#Код:
random_array = np.round(np.random.rand(10) * 10)
print(f"Изначальный вектор: \n {random_array} \n")
print(f"Наибольшее значение в данном векторе: {np.max(random_array)} \n")
random_array[(random_array == np.max(random_array))] = 0
print(f"Измененный вектор: \n {random_array}")

Изначальный вектор: 
 [1. 2. 3. 7. 9. 8. 5. 5. 5. 2.] 

Наибольшее значение в данном векторе: 9.0 

Измененный вектор: 
 [1. 2. 3. 7. 0. 8. 5. 5. 5. 2.]
